Python/Product/TestAdapter/PythonRunSettings.cs

```diff
@@ -0,0 +1,227 @@
+            _dispatcher = Dispatcher.CurrentDispatcher;
+            _serviceProvider = serviceProvider;
+        }
+
+        private void StateChange(object sender, OperationStateChangedEventArgs e) {
+            if (e.State == TestOperationStates.TestExecutionFinished) {
+                var resultUris = e.Operation.GetRunSettingsDataCollectorResultUri(PythonCodeCoverageUri);
+                if (resultUris != null) {
+                    foreach (var eachAttachment in resultUris) {
+                        string filePath = eachAttachment.LocalPath;
+
+                        if (!string.IsNullOrEmpty(filePath) && File.Exists(filePath)) {
+                            object inObj = filePath;
+                            object outObj = null;
+
+                            var dte = (EnvDTE.DTE)_serviceProvider.GetService(typeof(EnvDTE.DTE));
+                            dte.Commands.Raise(
+                                GuidList.guidPythonToolsCmdSet.ToString("B"),
+                                cmdidImportCoverage,
+                                ref inObj,
+                                ref outObj
+                            );
+                        }
+                    }
+                }
+            }
+        }
+
+        public ICodeCoverageSettingsService CodeCoverage {
+            get {
+                return _compModel.GetService<ICodeCoverageSettingsService>();
+            }
+        }
+
+        public string Name {
+            get {
+                return "Python Run Settings";
+            }
+        }
+
+        public IXPathNavigable AddRunSettings(IXPathNavigable inputRunSettingDocument, IRunSettingsConfigurationInfo configurationInfo, ILogger log) {
+            XPathNavigator navigator = inputRunSettingDocument.CreateNavigator();
+            var python = navigator.Select("/RunSettings");
+            if (python.MoveNext()) {
+                using (var writer = python.Current.AppendChild()) {
+                    var pyContainers = configurationInfo.TestContainers
+                        .Where(x => x is TestContainer)
+                        .Cast<TestContainer>()
+                        .GroupBy(x => x.Project);
+
+                    writer.WriteStartElement("Python");
+                    writer.WriteStartElement("TestCases");
+
+                    foreach (var project in pyContainers) {
+                        foreach (var container in project) {
+                            writer.WriteStartElement("Project");
+                            writer.WriteAttributeString("home", project.Key.ProjectHome);
+
+                            LaunchConfiguration config = null;
+                            string nativeCode = "", djangoSettings = "";
+                            _dispatcher.Invoke(() => {
+                                try {
+                                    config = project.Key.GetLaunchConfigurationOrThrow();
+                                } catch {
```

Comment:
If this happens, we're just going to crash below. We need to do something much more sensible when we can't get a launch configuration from the project.